packages/interactivity/src/proxies/test/state-proxy.ts

```diff
@@ -0,0 +1,1187 @@
+				},
+			} );
+
+			expect( state.y ).toEqual( [ 1, 2 ] );
+		} );
+
+		it( 'should support getters accessing the scope', () => {
+			const state = proxifyState( 'test', {
+				get y() {
+					const ctx = getContext< { value: string } >();
+					return ctx.value;
+				},
+			} );
+
+			const scope = { context: { test: { value: 'from context' } } };
+			try {
+				setScope( scope as any );
+				setNamespace( 'test' );
+				expect( state.y ).toBe( 'from context' );
+			} finally {
+				resetNamespace();
+				resetScope();
+			}
+		} );
+
+		it( 'should work with normal functions', () => {
+			const state = proxifyState( 'test', {
+				value: 1,
+				isBigger: ( newValue: number ): boolean =>
+					state.value < newValue,
+				sum( newValue: number ): number {
+					return state.value + newValue;
+				},
+				replace: ( newValue: number ): void => {
+					state.value = newValue;
+				},
+			} );
+			expect( state.isBigger( 2 ) ).toBe( true );
+			expect( state.sum( 2 ) ).toBe( 3 );
+			expect( state.value ).toBe( 1 );
+			state.replace( 2 );
+			expect( state.value ).toBe( 2 );
+		} );
+
+		it( 'should work with normal functions accessing the scope', () => {
+			const state = proxifyState( 'test', {
+				sumContextValue( newValue: number ): number {
+					const ctx = getContext< { value: number } >();
+					return ctx.value + newValue;
+				},
+			} );
+
+			const scope = { context: { test: { value: 1 } } };
+			try {
+				setScope( scope as any );
+				setNamespace( 'test' );
+				expect( state.sumContextValue( 2 ) ).toBe( 3 );
+			} finally {
+				resetNamespace();
+				resetScope();
+			}
+		} );
+
+		it( 'should allow using `this` inside functions', () => {
+			const state = proxifyState( 'test', {
+				value: 1,
+				sum( newValue: number ): number {
+					return this.value + newValue;
+				},
+			} );
+			expect( state.sum( 2 ) ).toBe( 3 );
+		} );
+	} );
+
+	describe( 'set', () => {
+		it( 'should update like plain objects/arrays', () => {
+			expect( state.a ).toBe( 1 );
+			expect( state.nested.b ).toBe( 2 );
+			state.a = 2;
+			state.nested.b = 3;
+			expect( state.a ).toBe( 2 );
+			expect( state.nested.b ).toBe( 3 );
+		} );
+
+		it( 'should support setting values with setters', () => {
+			const state = proxifyState( 'test', {
+				counter: 1,
+				get double() {
+					return state.counter * 2;
+				},
+				set double( val ) {
+					state.counter = val / 2;
+				},
+			} );
+			expect( state.counter ).toBe( 1 );
+			state.double = 4;
+			expect( state.counter ).toBe( 2 );
+		} );
+
+		it( 'should update array length', () => {
+			expect( state.array.length ).toBe( 2 );
+			state.array.push( 4 );
+			expect( state.array.length ).toBe( 3 );
+			state.array.splice( 1, 2 );
+			expect( state.array.length ).toBe( 1 );
+		} );
+
+		it( 'should update when mutations happen', () => {
+			expect( state.a ).toBe( 1 );
+			state.a = 11;
+			expect( state.a ).toBe( 11 );
+		} );
+
+		it( 'should support setting getters on the fly', () => {
```

Comment:
Maybe we should add a test similar to this one, but where `state.double` does exist (it's not `undefined`).